File: backend/app/schemas/activity.py
import uuid
from datetime import datetime
from typing import Optional, Any
from pydantic import BaseModel

class UserMiniResponse(BaseModel):
    id: uuid.UUID
    email: str
    full_name: str
    profile_image: Optional[str] = None

    class Config:
        from_attributes = True

class ActivityResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    category: Optional[str] = None
    event_type: Optional[str] = None
    action: str
    details: str
    target_type: Optional[str] = None
    target_name: Optional[str] = None
    target_id: Optional[uuid.UUID] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    meta_data: Optional[dict[str, Any]] = None
    created_at: datetime
    user: UserMiniResponse

    class Config:
        from_attributes = True

class AuditLogFilter(BaseModel):
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    user_id: Optional[uuid.UUID] = None
    category: Optional[str] = None
    event_type: Optional[str] = None
    target_type: Optional[str] = None
    action: Optional[str] = None

File: frontend/services/activity.ts
import { api } from './api'

export interface UserMini {
  id: string
  email: string
  full_name: string
  profile_image: string | null
}

export interface ActivityResponse {
  id: string
  user_id: string
  category: string | null
  event_type: string | null
  action: string
  details: string
  target_type: string | null
  target_name: string | null
  target_id: string | null
  ip_address: string | null
  user_agent: string | null
  created_at: string
  user: UserMini
}

export interface AuditStats {
  total_logs: number
  category_counts: Record<string, number>
  event_type_counts: Record<string, number>
}

export const activityService = {
  /**
   * Load activity timeline logs with scrolling pagination.
   */
  async getActivities(params?: {
    limit?: number
    offset?: number
    user_id?: string
    target_type?: string
  }): Promise<ActivityResponse[]> {
    const response = await api.get<ActivityResponse[]>('/activities', { params })
    return response.data
  },

  /**
   * Get filtered audit logs with comprehensive filtering.
   */
  async getAuditLogs(params?: {
    start_date?: string
    end_date?: string
    user_id?: string
    category?: string
    event_type?: string
    target_type?: string
    action?: string
    limit?: number
    offset?: number
  }): Promise<ActivityResponse[]> {
    const response = await api.get<ActivityResponse[]>('/activities/audit', { params })
    return response.data
  },

  /**
   * Get audit log statistics.
   */
  async getAuditStats(params?: {
    start_date?: string
    end_date?: string
  }): Promise<AuditStats> {
    const response = await api.get<AuditStats>('/activities/audit/stats', { params })
    return response.data
  },

  /**
   * Export audit logs to CSV format.
   */
  async exportAuditLogs(params?: {
    start_date?: string
    end_date?: string
    user_id?: string
    category?: string
    event_type?: string
    target_type?: string
    action?: string
  }): Promise<Blob> {
    const response = await api.get('/activities/audit/export', {
      params,
      responseType: 'blob'
    })
    return response.data
  }
}

